File: src/C/bstNull/bst.c
/* Immutable binary search tree in C
 *  by Atsushi Igarashi
 *  on August 31, 2016
 */

#include <stdio.h>   // for printf
#include <stdlib.h>  // for malloc
#include <stdbool.h> // for type bool, true, and false

struct tree {
  struct tree *left;
  int value;
  struct tree *right;
};

struct tree *newbranch(struct tree *_left,
                       int _value,
                       struct tree *_right) {
  // Allocate a new object in the heap
  struct tree *n = (struct tree *)malloc(sizeof(struct tree));
  // And then initialize the members
  n->left = _left;
  n->value = _value;
  n->right = _right;
  return n;
}

struct tree *newleaf() {
  /* Real C programmers would avoid defining such a simple function.
   * It causes overhead of function calls.
   */
  return NULL;
}

bool find(struct tree *t, int n) {
  if (t == NULL) {
    return false;
  } else /* t is a branch */ {
    if (n == t->value) {
      return true;
    } else if (n < t->value) {
      return find(t->left, n);
    } else /* n > t->value */ {
      return find(t->right, n);
    }
  }
}

struct tree *insert(struct tree *t, int n) {
  if (t == NULL) {
    return newbranch(newleaf(), n, newleaf());
  } else /* t is a branch */ {
    if (n == t->value) {
      return t;
    } else if (n < t->value) {
      return newbranch(insert(t->left, n), t->value, t->right);
    } else /* n > t->value */ {
      return newbranch(t->right, t->value, insert(t->right, n));
    }
  }
}

int min(struct tree *t) {
  if (t == NULL) {
    return -255;
  } else /* t is a branch */ {
    if (t->left == NULL) {
      return t->value;
    } else {
      return min(t->left);
    }
  }
}

struct tree *delete(struct tree *t, int n) {
  if (t == NULL) {
    return t;
  } else /* t is a branch */ {
    if (n == t->value) {
      if (t->left == NULL) {
        if (t->right == NULL) {
          return newleaf();
        } else /* t->right is a branch */ {
          return t->right;
        }
      } else /* t->left is a branch */ {
        if (t->right == NULL) {
          return t->left;
        } else /* t->right is a branch */ {
          int m = min(t->right);
          struct tree *newRight = delete(t->right, m);
          return newbranch(t->left, m, newRight);
        }
      }
    } else if (n < t->value) {
      struct tree *newLeft = delete(t->left, n);
      return newbranch(newLeft, t->value, t->right);
    } else /* n > t->value */ {
      struct tree *newRight = delete(t->right, n);
      return newbranch(t->left, t->value, newRight);
    }
  }
}

int main() {
  struct tree *t1 = newbranch(newleaf(), 10, newleaf());
  struct tree *t2 = newbranch(newleaf(), 25, newleaf());
  struct tree *t3 = newbranch(t1, 15, t2);
  struct tree *t4 = newbranch(newleaf(), 60, newleaf());
  struct tree *t5 = newbranch(newleaf(), 48, t4);
  struct tree *t6 = newbranch(t3, 30, t5);
  bool test1 = find(t6, 30);  // should be true
  bool test2 = find(t6, 13);  // should be false
  struct tree *t7 = insert(t6, 23);
  struct tree *t8 = insert(t6, 0);
  bool test3 = find(t7, 23);  // should be true
  bool test4 = find(t8, 30);  // should be true
  bool test5 = find(t8, 23);  // should be false
  struct tree *t9 = delete(t8, 30);
  bool test6 = find(t9, 30);  // should be false
  bool test7 = find(t9,48);  // should be true
  printf("test 1: %d\n", test1);
  printf("test 2: %d\n", test2);
  printf("test 3: %d\n", test3);
  printf("test 4: %d\n", test4);
  printf("test 5: %d\n", test5);
  printf("test 6: %d\n", test6);
  printf("test 7: %d\n", test7);
  return 0;
}

File: src/ocaml/bstModule/bst.ml
(* Works only in OCaml 4.03.0 or above *)

module type BST =
  sig
    type t
    val empty : t
    val find : t * int -> bool
    val insert : t * int -> t
    val delete : t * int -> t
  end

module Bst : BST =
  struct
    (* Defining the shape of trees *)
    type t =
      Lf       (* Leaf *)
    | Br of {  (* Branch *)
        left:  t;
        value: int;
        right: t;
      }

    let empty = Lf

    (* (Recursive) function find, which returns whether given integer n exists in BST t *)
    let rec find(t, n) =
      match t with
        Lf -> false
      | Br {left=l; value=v; right=r} ->
         if n = v then true
         else if n < v then find(l, n)
         else (* n > v *) find(r, n)

    (* (Recursive) function insert, which, given BST t and a new element n, returns
   a new binary search tree with n *)
    let rec insert(t, n) =
      match t with
        Lf -> Br {left=Lf; value=n; right=Lf}
      | Br {left=l; value=v; right=r} ->
         if n = v then t
         else if n < v then Br {left=insert(l, n); value=v; right=r}
         else (* n > v *) Br {left=l; value=v; right=insert(r, n)}

    (* Function min, which, given BST t, returns the minimum value stored in t.
   If t is empty, it returns -255. *)
    let rec min t =
      match t with
        Lf -> -255
      | Br {left=Lf; value=v; right=_} -> v
      | Br {left=l; value=_; right=_} -> min l

    (* (Recursive) function delete, which, given BST t and an element n to
   be deleted, returns a new binary search tree without n.  If n is not
   stored in t, it returns t as it is. *)

    let rec delete(t, n) =
      match t with
        Lf -> t
      | Br {left=l; value=v; right=r} ->
         if n = v then
           match l, r with
	     Lf, Lf -> Lf
           | Br _, Lf -> l
           | Lf, Br _ -> r
           | Br _, Br _ ->
	      let m = min r in
	      Br {left=l; value=m; right=delete(r, m)}
         else if n < v then Br {left=delete(l, n); value=v; right=r}
         else (* n > v *) Br {left=l; value=v; right=delete(r, n)}
end

(* Constructing a sample tree *)
let t0 = Bst.empty
let t1 = Bst.insert(t0, 30)
let t2 = Bst.insert(t1, 15)
let t3 = Bst.insert(t2, 48)
let t4 = Bst.insert(t3, 10)
let t5 = Bst.insert(t4, 25)
let t6 = Bst.insert(t5, 60)

open Bst

(* Testing find *)
let test1 = find(t6, 30)  (* should be true *)
let test2 = find(t6, 13)  (* should be false *)

(* Testing insert *)
let t7 = insert(t6, 23)
let t8 = insert(t6, 0)
let test3 = find(t7, 23)  (* should return true *)
let test4 = find(t8, 30)  (* should return false *)
let test5 = find(t8, 23)  (* should return false *)

(* Testing delete *)
let t9 = delete(t8, 30)
let test6 = find(t9, 30)
let test7 = find(t9, 48)
